File: src/npu_webhook/api/ingest.py
"""POST /ingest - 知识注入"""

from fastapi import APIRouter, HTTPException

from npu_webhook.app_state import state
from npu_webhook.models.schemas import IngestRequest, IngestResponse

router = APIRouter(prefix="/api/v1", tags=["ingest"])


@router.post("/ingest", response_model=IngestResponse)
async def ingest(req: IngestRequest) -> IngestResponse:
    """接收浏览器推送的内容（对话/网页/选中文本）"""
    if not state.db:
        raise HTTPException(status_code=503, detail="Database not initialized")

    from npu_webhook.config import settings

    content_len = len(req.content)
    if content_len < settings.ingest.min_content_length:
        raise HTTPException(
            status_code=400,
            detail=f"Content too short (min {settings.ingest.min_content_length} chars)",
        )
    # 防止超大内容导致 embedding 内存溢出（上限 500KB 字符）
    MAX_CONTENT = 500_000
    if content_len > MAX_CONTENT:
        raise HTTPException(
            status_code=413,
            detail=f"Content too large (max {MAX_CONTENT} chars, got {content_len})",
        )

    if req.domain and req.domain in settings.ingest.excluded_domains:
        raise HTTPException(status_code=400, detail="Domain is excluded")

    # 文本级近重复检测（前 200 字符匹配）
    existing_id = state.db.find_near_duplicate(req.content, req.source_type)
    if existing_id:
        return IngestResponse(id=existing_id, duplicate=True)

    # 存入 SQLite
    item_id = state.db.insert_item(
        title=req.title,
        content=req.content,
        source_type=req.source_type,
        url=req.url,
        domain=req.domain,
        tags=req.tags,
        metadata=req.metadata,
    )

    # 分块并投递 embedding 队列（P1 近实时）
    if state.chunker:
        sections = state.chunker.extract_sections(req.content, source_type=req.source_type)

        # Level 1: 章节
        for section_idx, section_text in sections:
            if section_text.strip():
                state.db.enqueue_embedding(
                    item_id=item_id,
                    chunk_index=section_idx,
                    chunk_text=section_text,
                    priority=1,
                    level=1,
                    section_idx=section_idx,
                )

        # Level 2: 段落块
        chunk_counter = 0
        for section_idx, section_text in sections:
            chunks = state.chunker.chunk(section_text)
            for chunk_text in chunks:
                state.db.enqueue_embedding(
                    item_id=item_id,
                    chunk_index=chunk_counter,
                    chunk_text=chunk_text,
                    priority=1,
                    level=2,
                    section_idx=section_idx,
                )
                chunk_counter += 1

    return IngestResponse(id=item_id)

File: npu-vault/crates/vault-server/src/routes/ingest.rs
use axum::extract::State;
use axum::http::StatusCode;
use axum::Json;
use serde::Deserialize;
use vault_core::chunker;

use crate::state::SharedState;

#[derive(Deserialize)]
pub struct IngestRequest {
    pub title: String,
    pub content: String,
    #[serde(default = "default_source_type")]
    pub source_type: String,
    pub url: Option<String>,
    pub domain: Option<String>,
    pub tags: Option<Vec<String>>,
}

fn default_source_type() -> String {
    "note".into()
}

/// JSON ingest 内容上限（防止大负载写放大攻击）
const MAX_INGEST_CONTENT: usize = 2 * 1024 * 1024; // 2 MB
const MAX_INGEST_TITLE: usize = 500;

pub async fn ingest(
    State(state): State<SharedState>,
    Json(body): Json<IngestRequest>,
) -> Result<Json<serde_json::Value>, (StatusCode, Json<serde_json::Value>)> {
    if body.title.len() > MAX_INGEST_TITLE {
        return Err((
            StatusCode::PAYLOAD_TOO_LARGE,
            Json(serde_json::json!({"error": format!("title too long (max {MAX_INGEST_TITLE} bytes)")})),
        ));
    }
    if body.content.len() > MAX_INGEST_CONTENT {
        return Err((
            StatusCode::PAYLOAD_TOO_LARGE,
            Json(serde_json::json!({"error": format!("content too large: {} bytes (max {MAX_INGEST_CONTENT})", body.content.len())})),
        ));
    }
    let vault = state.vault.lock()
        .map_err(|_| (StatusCode::INTERNAL_SERVER_ERROR, Json(serde_json::json!({"error": "vault lock poisoned"}))))?;
    let dek = vault.dek_db().map_err(|e| {
        (StatusCode::FORBIDDEN, Json(serde_json::json!({"error": e.to_string()})))
    })?;

    let id = vault
        .store()
        .insert_item(
            &dek,
            &body.title,
            &body.content,
            body.url.as_deref(),
            &body.source_type,
            body.domain.as_deref(),
            body.tags.as_deref(),
        )
        .map_err(|e| {
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(serde_json::json!({"error": e.to_string()})),
            )
        })?;

    // Invalidate search cache after new item inserted
    {
        if let Ok(mut cache) = state.search_cache.lock() {
            cache.clear();
        }
    }

    // Add to fulltext index
    {
        let ft_guard = state.fulltext.lock()
            .map_err(|_| (StatusCode::INTERNAL_SERVER_ERROR, Json(serde_json::json!({"error": "ft lock poisoned"}))))?;
        if let Some(ft) = ft_guard.as_ref() {
            let _ = ft.add_document(&id, &body.title, &body.content, &body.source_type);
        }
    }

    // Enqueue for embedding: two-layer indexing (sections L1 + chunks L2)
    {
        let sections = chunker::extract_sections(&body.content);
        let mut chunk_counter = 0;

        // Level 1: section-level embeddings
        for (section_idx, section_text) in &sections {
            if !section_text.trim().is_empty() {
                if let Err(e) = vault.store().enqueue_embedding(
                    &id, chunk_counter, section_text, 1, 1, *section_idx,
                ) {
                    tracing::warn!("enqueue_embedding L1 failed for item {id}: {e}");
                }
                chunk_counter += 1;
            }
        }

        // Level 2: paragraph chunk embeddings
        for (section_idx, section_text) in &sections {
            for chunk_text in
                chunker::chunk(section_text, chunker::DEFAULT_CHUNK_SIZE, chunker::DEFAULT_OVERLAP)
            {
                if let Err(e) = vault.store().enqueue_embedding(
                    &id, chunk_counter, &chunk_text, 2, 2, *section_idx,
                ) {
                    tracing::warn!("enqueue_embedding L2 failed for item {id}: {e}");
                }
                chunk_counter += 1;
            }
        }
    }

    // Auto-enqueue classification
    if let Err(e) = vault.store().enqueue_classify(&id, 3) {
        tracing::warn!("enqueue_classify failed for item {id}: {e}");
    }

    Ok(Json(serde_json::json!({
        "id": id,
        "status": "ok"
    })))
}
